os/net/routing/rpl-classic/rpl-nbr-policy.c

```diff
@@ -27,231 +28,112 @@
-    LOG_DBG("nbr-policy: num-free > 0 = %d - Other for RPL/ND6 unused NBR entry exists.\n",
-           num_free);
-  }
-  if(num_children < MAX_CHILDREN) {
-    return worst_rank_nbr;
-  }
-  return NULL;
+  return get_rank(lladdr2) > get_rank(lladdr1) ? lladdr2 : lladdr1;
 }
 /*---------------------------------------------------------------------------*/
-const linkaddr_t *
-find_removable_dio(uip_ipaddr_t *from, rpl_dio_t *dio)
+static bool
+can_accept_new_parent(const linkaddr_t *candidate_for_removal, rpl_dio_t *dio)
 {
-  rpl_instance_t *instance;
-
-  update_nbr();
-
-  instance = rpl_get_instance(dio->instance_id);
-  if(instance == NULL || instance->current_dag == NULL) {
-    LOG_WARN("nbr-policy: did not find instance id: %d\n", dio->instance_id);
-    return NULL;
-  }
-
-  /* Add the new neighbor only if it is better than the worst parent. */
-  if(dio->rank + instance->min_hoprankinc < worst_rank - instance->min_hoprankinc / 2) {
-    /* Found *great* neighbor - add! */
-    LOG_DBG("nbr-policy: DIO rank %u, worst_rank %u -- add to cache\n",
-           dio->rank, worst_rank);
-
-    return worst_rank_nbr;
-  }
-
-  LOG_DBG("nbr-policy: DIO rank %u, worst_rank %u -- do not add to cache\n",
-         dio->rank, worst_rank);
-  return NULL;
-}
-/*---------------------------------------------------------------------------*/
-const linkaddr_t *
-find_removable_dao(uip_ipaddr_t *from, rpl_instance_t *instance)
-{
-  int max = MAX_CHILDREN;
-  update_nbr();
-
-  if(instance != NULL) {
-    /* No need to reserve space for parents for RPL ROOT */
-    if(instance->current_dag->rank == ROOT_RANK(instance)) {
-      max = NBR_TABLE_MAX_NEIGHBORS;
+  rpl_rank_t rank_candidate;
+  /* There's space left in the table or the worst entry has no rank: accept */
+  if(candidate_for_removal == NULL
+     || (rank_candidate = get_rank(candidate_for_removal)) == RPL_INFINITE_RANK) {
+    return true;
+  } else {
+    rpl_instance_t *instance = rpl_get_default_instance();
+    rpl_rank_t new_path_rank;
+
+    if(instance == NULL || dio == NULL) {
+      return false;
     }
-  }
 
-  /* Check if this DAO sender is not yet neighbor and there is already too
-     many children. */
-  if(num_children >= max) {
-    LOG_ERR("nbr-policy: can not add another child - already at max.\n");
-    return NULL;
+    new_path_rank = dio->rank + instance->min_hoprankinc;
+    return new_path_rank < rank_candidate - instance->min_hoprankinc / 2;
   }
-  /* remove the worst ranked nbr */
-  return worst_rank_nbr;
 }
 /*---------------------------------------------------------------------------*/
-const linkaddr_t *
-rpl_nbr_policy_find_removable(nbr_table_reason_t reason,void * data)
+bool
+rpl_nbr_can_accept_new(const linkaddr_t *new, const linkaddr_t *candidate_for_removal,
+                       nbr_table_reason_t reason, void *data)
 {
-  /* When we get the DIO/DAO/DIS we know that UIP contains the
-     incoming packet */
+  bool accept;
   switch(reason) {
```

Comment:
Shouldn't there be breaks in the cases where accept is set?

Reply:
Oops well yes, absolutely. Done